### ISSUE TITLE
Editing White Box component shapes sometimes causes an Editor crash

### ISSUE THREAD

**Issue by jckand-amzn** [o3de/o3de#2702]
**Describe the bug**
Editing White Box component shapes sometimes causes an Editor crash.

**To Reproduce**
1. Create a new entity.
2. Add a White Box component to the entity.
3. Click Edit on the component.
4. Click and drag on one of the shape's faces/vertices/edges a few times while repositioning the main Editor camera.

**Expected behavior**
Shape can be edited without issue.

**Observed behavior**
Editor sometimes crashes. There appear to be a few different crash logs for what appears to be the same crash to the user.

**Stack**
Call Stack Trace:
25) 00007FF82B57C247 (KERNELBASE) : UnhandledExceptionFilter
24) 00007FF82DC35270 (ntdll) : memset
23) 00007FF82DC1C776 (ntdll) : _C_specific_handler
22) 00007FF82DC3217F (ntdll) : _chkstk
21) 00007FF82DBE1454 (ntdll) : RtlRaiseException
20) 00007FF82DC30CAE (ntdll) : KiUserExceptionDispatcher
19) 00007FF8185E15F4 (VCRUNTIME140) : memcpy
18) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\std\containers\vector.h (170) : AZStd::vector<WhiteBox::GenericHandle<WhiteBox::Api::HalfedgeHandleTag>,AZStd::allocator>::vector<WhiteBox::GenericHandle<WhiteBox::Api::HalfedgeHandleTag>,AZStd::allocator>
17) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Viewport\WhiteBoxPolygonScaleModifier.cpp (29) : WhiteBox::PolygonScaleModifier::PolygonScaleModifier
16) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\SubComponentModes\EditorWhiteBoxDefaultMode.cpp (462) : WhiteBox::DefaultMode::CreatePolygonScaleModifier
15) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<WhiteBox::EditorWhiteBoxDefaultModeRequests,WhiteBox::EditorWhiteBoxDefaultModeRequests,1,1>::Dispatcher<AZ::EBus<WhiteBox::EditorWhiteBoxDefault
14) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (175) : AZStd::Internal::function_util::get_invoker<void __cdecl(AzToolsFramework::LinearManipulator::Action const &),AZStd::Internal::function_util::function_obj_tag,AZStd::all
13) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Manipulators\LinearManipulator.cpp (194) : AzToolsFramework::LinearManipulator::OnLeftMouseUpImpl
12) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Manipulators\BaseManipulator.cpp (95) : AzToolsFramework::BaseManipulator::OnLeftMouseUp
11) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Manipulators\ManipulatorManager.cpp (220) : AzToolsFramework::ManipulatorManager::ConsumeViewportMouseRelease
10) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (823) : AZ::Internal::EBusContainer<AzToolsFramework::EditorInteractionSystemViewportSelection,AzToolsFramework::EditorInteractionSystemViewportSelectionRequests,1,0>::Dispatcher<AZ
 9) D:\GitHub\o3de\Code\Editor\ViewportManipulatorController.cpp (221) : SandboxEditor::ViewportManipulatorControllerInstance::HandleInputChannelEvent
 8) D:\GitHub\o3de\Code\Framework\AzFramework\AzFramework\Viewport\ViewportControllerList.cpp (66) : AzFramework::ViewportControllerList::HandleInputChannelEvent
 7) D:\Perforce\lyengine\3rdParty\packages\qt-5.15.2-rev4-windows\qt\include\QtCore\qobjectdefs_impl.h (443) : QtPrivate::QFunctorSlotObject<`AtomToolsFramework::RenderViewportWidget::InitializeViewportContext'::`2'::<lambda_1>,2,QtPrivate::List<AzFramework:
 6) 00007FFFD19B0DB9 (Qt5Core) : QObject::qt_static_metacall
 5) D:\GitHub\o3de\windows_vs2019\Code\Framework\AzToolsFramework\AzToolsFramework_autogen\include_profile\JQ32PZHRHS\moc_QtEventToAzInputManager.cpp (139) : AzToolsFramework::QtEventToAzInputMapper::InputChannelUpdated
 4) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Input\QtEventToAzInputManager.cpp (284) : AzToolsFramework::QtEventToAzInputMapper::eventFilter
 3) 00007FFFD1993AB2 (Qt5Core) : QCoreApplicationPrivate::sendThroughApplicationEventFilters
 2) 00007FFFCCFA48F2 (Qt5Widgets) : QApplicationPrivate::notify_helper
 1) 00007FFFCCFA3A33 (Qt5Widgets) : QApplication::notify

Call Stack Trace:
25) 00007FF82B57C247 (KERNELBASE) : UnhandledExceptionFilter
24) 00007FF82DC35270 (ntdll) : memset
23) 00007FF82DC1C776 (ntdll) : _C_specific_handler
22) 00007FF82DC3217F (ntdll) : _chkstk
21) 00007FF82DBE1454 (ntdll) : RtlRaiseException
20) 00007FF82DC30CAE (ntdll) : KiUserExceptionDispatcher
19) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Core\WhiteBoxToolApi.cpp (746) : WhiteBox::Api::FaceHalfedgeHandles
18) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Core\WhiteBoxToolApi.cpp (782) : WhiteBox::Api::FaceVertexHandles
17) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Core\WhiteBoxToolApi.cpp (980) : WhiteBox::Api::BorderHalfedgeHandles
16) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Core\WhiteBoxToolApi.cpp (1743) : WhiteBox::Api::PolygonBorderVertexHandles
15) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Viewport\WhiteBoxPolygonScaleModifier.cpp (61) : WhiteBox::PolygonScaleModifier::CreateManipulators
14) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\Viewport\WhiteBoxPolygonScaleModifier.cpp (31) : WhiteBox::PolygonScaleModifier::PolygonScaleModifier
13) D:\GitHub\o3de\Gems\WhiteBox\Code\Source\SubComponentModes\EditorWhiteBoxDefaultMode.cpp (462) : WhiteBox::DefaultMode::CreatePolygonScaleModifier
12) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<WhiteBox::EditorWhiteBoxDefaultModeRequests,WhiteBox::EditorWhiteBoxDefaultModeRequests,1,1>::Dispatcher<AZ::EBus<WhiteBox::EditorWhiteBoxDefault
11) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (175) : AZStd::Internal::function_util::get_invoker<void __cdecl(AzToolsFramework::LinearManipulator::Action const &),AZStd::Internal::function_util::function_obj_tag,AZStd::all
10) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Manipulators\LinearManipulator.cpp (194) : AzToolsFramework::LinearManipulator::OnLeftMouseUpImpl
 9) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Manipulators\BaseManipulator.cpp (95) : AzToolsFramework::BaseManipulator::OnLeftMouseUp
 8) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Manipulators\ManipulatorManager.cpp (220) : AzToolsFramework::ManipulatorManager::ConsumeViewportMouseRelease
 7) D:\GitHub\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (823) : AZ::Internal::EBusContainer<AzToolsFramework::EditorInteractionSystemViewportSelection,AzToolsFramework::EditorInteractionSystemViewportSelectionRequests,1,0>::Dispatcher<AZ
 6) D:\GitHub\o3de\Code\Editor\ViewportManipulatorController.cpp (221) : SandboxEditor::ViewportManipulatorControllerInstance::HandleInputChannelEvent
 5) D:\GitHub\o3de\Code\Framework\AzFramework\AzFramework\Viewport\ViewportControllerList.cpp (66) : AzFramework::ViewportControllerList::HandleInputChannelEvent
 4) D:\Perforce\lyengine\3rdParty\packages\qt-5.15.2-rev4-windows\qt\include\QtCore\qobjectdefs_impl.h (443) : QtPrivate::QFunctorSlotObject<`AtomToolsFramework::RenderViewportWidget::InitializeViewportContext'::`2'::<lambda_1>,2,QtPrivate::List<AzFramework:
 3) 00007FFFD19B0DB9 (Qt5Core) : QObject::qt_static_metacall
 2) D:\GitHub\o3de\windows_vs2019\Code\Framework\AzToolsFramework\AzToolsFramework_autogen\include_profile\JQ32PZHRHS\moc_QtEventToAzInputManager.cpp (139) : AzToolsFramework::QtEventToAzInputMapper::InputChannelUpdated
 1) D:\GitHub\o3de\Code\Framework\AzToolsFramework\AzToolsFramework\Input\QtEventToAzInputManager.cpp (284) : AzToolsFramework::QtEventToAzInputMapper::eventFilter

**Comment by willihay** [comment]
I experienced this too. I was able to get a consistent crash when editing a white box cylinder, in particular:

1. Go into white box edit mode.
2. Hover over cylinder vertice until red dot appears.
3. Hold Ctrl key down.
4. Drag dot up.

-> crash

**Comment by hultonha** [comment]
Blast... thanks for the repro steps, I'll try and get to this soon! 👍

**Comment by hultonha** [comment]
I've started looking into this and can see it's something related to viewport input (not just White Box). It's likely White Box should be able to handle this more gracefully, but it's not actually the source of the bug.

At the time of writing it _seems_ to be related to how viewport events are handled (and propagated) and this causes problems with Manipulators which in turn causes problems with White Box.

I'll update this ticket when I have more info...

**Comment by hultonha** [comment]
Okay I've tracked down the problem - it's related to this line in `ViewportManipulatorController`

```c++
return interactionHandled && event.m_inputChannel.IsActive();
```

The problematic part is `event.m_inputChannel.IsActive()` - what can happen is with a move event, the input channel can be Idle (to `Began` or `Active`) so after a manipulator move event, instead of stopping the event propagating, the same move event will get sent to the Viewport as well (which shouldn't happen), and this then leads to logic that should happen when no manipulator is active (which in-turn causes a crash).

Now I'm not sure why `event.m_inputChannel.IsActive()` was added but I believe it should be removed (or at least updated to handle the channel being Idle). @NicholasVanSickle I think knows best why this is the case, could you let me know why this additional check is needed (instead of just handled?).

I am going to try and write a test to verify this problem and ensure the correct behavior is observed after the fix. @NicholasVanSickle are there any relevant tests that already exist to confirm the current behavior? If this isn't the case is there a manual process I can try to ensure things do the right thing? (We really should try and add a test to catch whatever breaks if `event.m_inputChannel.IsActive()` is removed if it really is needed.

**Comment by hultonha** [comment]
I've nearly got a test in place to verify this (which should be helpful in future for more ViewportManipulatorController tests) but the puzzling thing is we apparently do not forward events if they're idle, so I'm not sure how the `m_inputChannel` state being idle can ever get there. Will see if I can investigate what is causing this.

**Comment by hultonha** [comment]
Fixed by #3210
